### ISSUE TITLE
Proposal add `docker images --tree` back

### ISSUE THREAD

**Issue by HuKeping** [docker/docker#17366]
Should we add this option back since a lot of people want it? 

https://github.com/docker/docker/pull/5001#issuecomment-151058939

**Comment by fake-name** [comment]
Yes.

**Comment by bragboy** [comment]
+1

**Comment by avinb** [comment]
yes

**Comment by joernhees** [comment]
:+1:

i understand and fully support the decision to remove fancy cruft from the core of tools... (like generating nice pictures and dependency graphs in this case)

However, the ascii only version `docker images --tree` doesn't seem like a high maintenance burden to me. By cutting it out of the core it seems like you cut away each admin's ability to quickly understand what's going on with containers, how they depend on each other etc.

please bring it back

**Comment by julian1** [comment]
yes please!

**Comment by kunalkushwaha** [comment]
+1

**Comment by stevvooe** [comment]
Let's revisit [history](https://github.com/docker/docker/pull/5001#issue-30818869):

> As discussed on #docker-dev, these flags can easily be implemented as external tools.

Are we not making enough information available via the cli or api to make tools like this possible?

**Comment by joernhees** [comment]
There are external tools like https://github.com/justone/dockviz , so the information seems to be available somehow, just that there doesn't seem to be a single human understandable built-in anymore (please correct me if i'm wrong). As docker's image layers form a tree, it seems wrong to not have a single built-in non-fancy terminal only visualization for admins to understand that tree.

As developers of a tool you obviously have to decide what to include in core and what not. I think the amount of feedback from your community about removing `docker images --tree` indicates that you didn't just cut away a "nice to have" feature, but more an essential and easy to understand visualization of the "what is basing on what" use-case.

How complicated would it be to re-introduce this? If it's not much code and doesn't pull in any external dependencies, why not just give us that back? (Note how this is not about more fancy stuff like `--viz`!)

Let me add that especially for newcomers to docker such a visualization explains a lot. Requiring them to first install some external tool with docker before understanding docker is a bit weird...

**Comment by bragboy** [comment]
Totally agree with @joernhees

**Comment by fake-name** [comment]
> Let's revisit history:

>> As discussed on #docker-dev, these flags can easily be implemented as external tools.

1. Ok, great, somewhere, in a IRC channel at some point in time, some people talked about something. Congradulations. That's completely useless as a reference.
2. They can easily be implemented as external tools:
    - Fine, that's wonderful. **Implement them externally, and *then* remove the internal implementation**.

Right now, the "solution" to this issue is to track down some guy who created a freaking *entire image* to do this, *download the **several hundred megabyte OS image***, and keep that around *just* for the purpose of implementing a command that probably required ~1-10K of underlying source code in the actual docker tool.

So basically, you're claiming there are alternatives, but that's like claiming a F1 dragster is an alternative to a bicycle. Sure, they both get you from place A to place B, but the maintenance overhead of the former is enormous.

**Comment by stevvooe** [comment]
@fake-name I'm sorry if my original response was terse. I was trying to bring light to the original reasons for removing this feature, which are still valid today. I hope you can take the time to see this side of the argument.

Here is the whole quote:

> As discussed on #docker-dev, these flags can easily be implemented as external tools. We want to keep the Docker core as small as possible. Let's start cutting down the "nice-to-have" features.

> - The commands are no longer listed or documented.
> - The commands still work but print a deprecation warning.
> - The commands should be removed in a future version.

If you had clicked through that link you would also have seen see that "some guy" is @shykes, who started this project. His reasoning behind this removal was to keep the core of docker small.

Adding 1-10K lines of code is no small task. Keeping that code maintained over time, in the face of a rapidly evolving project, not only takes resources away from core features but adds extra burden to ensure those features don't break features like this. Check out #17924 if you're interested in an example of where something like this would impact ongoing features.

That said, if someone were to submit a PR, it would probably be considered.

**Comment by remram44** [comment]
This is a very important command that most users are using all the time. While I believe having extra non-essential complex commands as plugins make sense, it is definitely not the case of images --tree.

> That said, if someone were to submit a PR, it would probably be considered.

This is not enough of a statement for anyone to write a PR. Obviously.

**Comment by nigelpoulton** [comment]
Things like `docker images --tree` are vital for sys admins and operations teams at large enterprises. It's my personal opinion that overlooking basic features and functionality like this will hinder Docker adoption at such companies. 

I wrote about it briefly here - http://blog.nigelpoulton.com/savethetree-docker-in-enterprises/

But it remains my opinion that removing stuff like this is fine for developers and the likes, but absolutely not fine for people wanting to deploy Docker into live production estates that are guarded and prized like the crown jewels.

Please bring this back and please consider things like this in the future.

**Comment by adanielyan** [comment]
+1 for bringing --tree back

**Comment by tonistiigi** [comment]
After content addressability `--tree` isn't useful in its previous form. `--tree` shows image chains but with the new model all images are flat.

We could show something similar that visually shows how layers are shared between images. In a new model, every image has many layers and same layers can be shared between images. Atm all layer-chains have fixed order so this can be quite similar to a tree. The main difference is that we don't currently show layerID or even provide them through api. We have discussed something like `docker inspect --raw` for it.

So the first step would be to provide a way for the API to expose layers info. Then we can think what is the best visual way to show this info using a separate tool, and if this tool seems very useful we can discuss including it in docker cli client.

**Comment by tiborvass** [comment]
I agree with @stevvooe and @tonistiigi: I personally (and other maintainers too) don't think we should add `--tree` back, instead we should make sure the API provides ways of doing so. As @tonistiigi said, this today would need a new layers info API.

On the bright side, even if we wanted it back, that new layers info API would be the first step.

**Comment by duglin** [comment]
I'm wondering, and I've never used `--tree` so forgive me if this should be obvious, but do people want just the history/tree of how "images" related to each other or do they also want it to include the layers within images?

**Comment by HuKeping** [comment]
Anyway, have a new layer info API would be great.

**Comment by xificurC** [comment]
Since my issue was closed as a duplicate of this I'm gonna comment here - can we revive the discussion on this? Even `dockviz` isn't working anymore. Taking a (very short) glimpse into the code it seems it relied on the `Parent` field of images which I believe got pretty useless since `1.10`. The only way I see how one could build a tree is through the new `RootFS` layers in `docker inspect`, but that requires a bit of coding and more importantly relies on having the tagged images locally. Having the metadata where an image came from (i.e. the `FROM` clause of the `Dockerfile`) for **each** image would be of great benefit.

Thoughts?

**Comment by Karunamon** [comment]
Another workaround is [sen](https://github.com/TomasTomecek/sen), another (quite awesome) docker visualization tool, but that again carries the unfortunate burden of 57 megabytes of code and an inscrutable command to add something that was just fine before it was removed.

Please take this as another +1 by any means necessary. Knowing the structure of my images is not a "nice to have" as mentioned in #5001, it's a basic admin feature.

**Comment by TomasTomecek** [comment]
@Karunamon Thanks for advocating sen!

In latest release I polished the tree interface slightly:

![Tree view](https://raw.githubusercontent.com/TomasTomecek/sen/master/data/image-tree.gif)

Unfortunately, it almost doesn't work with newer versions of docker due to change of the data model (this was recorded with 1.9). With 1.11, `inspect` provides list of layers, but when you try to `inspect` them, engine responds with 404; therefore it's no longer possible to visualize layers of pulled images this way. Such pity. Fortunately it works with locally built images.

**Comment by justone** [comment]
Hey, I just released [dockviz v0.4](https://github.com/justone/dockviz/releases/tag/v0.4) that enables it to work with Docker versions higher than 1.9.x.  Please try it out and see if it works for you.

Turns out that using the layers in `docker inspect` was a non-starter because there are only layers for images that contain filesystem data changes.  So if you had two images that only differed at the end by metadata (like ENV, CMD, LABEL, etc.) then they would have the exact same set of layers in RootFS.  I ended up finding the solution in the API behind `docker history`.  Dockviz now synthesizes image IDs based on sha256 sums of the history data.  Works pretty well.

```
$ docker info | grep "Server Version"
Server Version: 1.11.1
$ dockviz03 images -t                                                                                                                                                                                                                                                                  
├─sha256:c18c7 Virtual Size: 577.1 MB Tags: skegio/clojure:java8
│ └─sha256:78e76 Virtual Size: 577.3 MB
│   └─sha256:a0d3a Virtual Size: 578.8 MB
│     └─sha256:02808 Virtual Size: 578.8 MB Tags: skeg-nate-20160429162403:latest
└─sha256:794ce Virtual Size: 444.2 MB Tags: skegio/base:latest
[master] nate@infinity:~/projects/mt/mwp-aws
$ dockviz04 images -t
└─<missing> Virtual Size: 120.1 MB
  └─<missing> Virtual Size: 120.1 MB
    └─<missing> Virtual Size: 120.1 MB
      └─<missing> Virtual Size: 120.1 MB
        └─<missing> Virtual Size: 120.1 MB
          └─<missing> Virtual Size: 120.1 MB
            └─<missing> Virtual Size: 120.1 MB
              └─<missing> Virtual Size: 268.2 MB
                └─<missing> Virtual Size: 442.6 MB
                  └─<missing> Virtual Size: 444.2 MB
                    └─<missing> Virtual Size: 444.2 MB
                      └─<missing> Virtual Size: 444.2 MB
                        └─<missing> Virtual Size: 444.2 MB
                          └─<missing> Virtual Size: 444.2 MB
                            └─<missing> Virtual Size: 444.2 MB
                              └─<missing> Virtual Size: 444.2 MB Tags: skegio/base:latest
                                └─<missing> Virtual Size: 575.8 MB
                                  └─<missing> Virtual Size: 575.8 MB
                                    └─<missing> Virtual Size: 575.8 MB
                                      └─sha256:c18c7 Virtual Size: 577.1 MB Tags: skegio/clojure:java8
                                        └─sha256:78e76 Virtual Size: 577.3 MB
                                          └─sha256:a0d3a Virtual Size: 578.8 MB
                                            └─sha256:02808 Virtual Size: 578.8 MB Tags: skeg-nate-20160429162403:latest
```

**Comment by LK4D4** [comment]
dockviz was added to docs, also sen is awesome too. way better than --tree was
